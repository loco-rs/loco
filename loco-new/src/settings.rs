//! Defines configurable application settings.

use std::env;

use heck::ToSnakeCase;
use rhai::{CustomType, TypeBuilder};
use serde::{Deserialize, Serialize};

use crate::{
    wizard::{self, AssetsOption, BackgroundOption, DBOption},
    LOCO_VERSION, OS,
};

/// Represents general application settings.
#[derive(Serialize, Deserialize, Clone, Debug, CustomType)]
pub struct Settings {
    pub package_name: String,
    pub module_name: String,
    pub db: Option<Db>,
    pub background: Option<Background>,
    pub asset: Option<Asset>,
    pub auth: bool,
    pub mailer: bool,
    pub clientside: bool,
    pub initializers: Option<Initializers>,
    pub features: Features,
    pub loco_version_text: String,
    pub os: OS,
}

impl From<DBOption> for Option<Db> {
    fn from(db_option: DBOption) -> Self {
        match db_option {
            DBOption::None => None,
            _ => Some(Db {
                kind: db_option.clone(),
                endpoint: db_option.endpoint_config().to_string(),
            }),
        }
    }
}

impl From<BackgroundOption> for Option<Background> {
    fn from(bg: BackgroundOption) -> Self {
        match bg {
            BackgroundOption::None => None,
            _ => Some(Background { kind: bg }),
        }
    }
}

impl From<AssetsOption> for Option<Asset> {
    fn from(asset: AssetsOption) -> Self {
        match asset {
            AssetsOption::None => None,
            _ => Some(Asset { kind: asset }),
        }
    }
}

impl Settings {
    /// Creates a new [`Settings`] instance based on prompt selections.
    #[must_use]
    pub fn from_wizard(package_name: &str, prompt_selection: &wizard::Selections, os: OS) -> Self {
        let features = if prompt_selection.db.enable() {
            Features::default()
        } else {
            let mut features = Features::disable_features();
            if prompt_selection.background.enable() {
                features.names.push("bg_redis".to_string());
            };
            features
        };

        Self {
            package_name: package_name.to_string(),
            module_name: package_name.to_snake_case(),
            auth: prompt_selection.db.enable(),
            mailer: prompt_selection.db.enable(),
            db: prompt_selection.db.clone().into(),
            background: prompt_selection.background.clone().into(),
            asset: prompt_selection.asset.clone().into(),
            clientside: prompt_selection.asset.enable(),
            initializers: if prompt_selection.asset.enable() {
                Some(Initializers { view_engine: true })
            } else {
                None
            },
            features,
            loco_version_text: get_loco_version_text(),
            os,
        }
    }
}
impl Default for Settings {
    fn default() -> Self {
        #[allow(clippy::default_trait_access)]
        Self {
            package_name: Default::default(),
            module_name: Default::default(),
            db: Default::default(),
            background: Default::default(),
            asset: Default::default(),
            auth: Default::default(),
            mailer: Default::default(),
            clientside: Default::default(),
            initializers: Default::default(),
            features: Default::default(),
            loco_version_text: get_loco_version_text(),
            os: Default::default(),
        }
    }
}

fn get_loco_version_text() -> String {
    env::var("LOCO_DEV_MODE_PATH").map_or_else(
        |_| format!(r#"version = "{LOCO_VERSION}""#),
        |path| {
            let path = path.replace('\\', "/");
            format!(r#"version="*", path="{path}""#)
        },
    )
}

/// Database configuration settings.
#[derive(Serialize, Deserialize, Clone, Debug, Default, CustomType)]
pub struct Db {
    pub kind: DBOption,
    pub endpoint: String,
}

/// Background processing configuration.
#[derive(Serialize, Deserialize, Clone, Debug, Default, CustomType)]
pub struct Background {
    pub kind: BackgroundOption,
}

/// Asset configuration settings.
#[derive(Serialize, Deserialize, Clone, Debug, Default, CustomType)]
pub struct Asset {
    pub kind: AssetsOption,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default, CustomType)]
pub struct Initializers {
    pub view_engine: bool,
}

/// Feature configuration, allowing toggling of optional features.
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Features {
    pub default_features: bool,
    pub names: Vec<String>,
}

impl Default for Features {
    fn default() -> Self {
        Self {
            default_features: true,
            names: vec![],
        }
    }
}

impl Features {
    /// Disables default features.
    #[must_use]
    pub fn disable_features() -> Self {
        Self {
            default_features: false,
            names: vec!["cli".to_string()],
        }
    }
}
