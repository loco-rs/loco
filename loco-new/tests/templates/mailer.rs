use loco::settings;
use rstest::rstest;

use super::*;
use crate::assertion;

pub fn run_generator(enable_mailer: bool) -> TestGenerator {
    let settings = settings::Settings {
        mailer: enable_mailer,
        ..Default::default()
    };

    TestGenerator::generate(settings)
}

#[rstest]
fn test_config_file_without_mailer(
    #[values("config/development.yaml", "config/test.yaml")] config_file: &str,
) {
    let generator = run_generator(false);
    let content = assertion::yaml::load(generator.path(config_file));
    assertion::yaml::assert_path_is_empty(&content, &["mailer"]);
}

#[rstest]
fn test_config_file_with_mailer(
    #[values("config/development.yaml", "config/test.yaml")] config_file: &str,
) {
    let generator = run_generator(true);
    let content = assertion::yaml::load(generator.path(config_file));
    assertion::yaml::assert_path_key_count(&content, &["mailer"], 1);
    assertion::yaml::assert_path_key_count(&content, &["mailer", "smtp"], 4);
    assertion::yaml::assert_path_value_eq_bool(&content, &["mailer", "smtp", "enable"], true);
    assertion::yaml::assert_path_value_eq_int(&content, &["mailer", "smtp", "port"], 1025);
    assertion::yaml::assert_path_value_eq_bool(&content, &["mailer", "smtp", "secure"], false);
    assertion::yaml::assert_path_value_eq_string(
        &content,
        &["mailer", "smtp", "host"],
        "localhost",
    );
}

#[rstest]
fn test_cargo_toml(#[values(true, false)] mailer: bool) {
    let generator = run_generator(mailer);
    let content = assertion::toml::load(generator.path("Cargo.toml"));

    insta::assert_snapshot!(
        format!("cargo_dependencies_mailer_{:?}", mailer),
        content.get("dependencies").unwrap()
    );
}

#[rstest]
fn test_src_lib_rs(#[values(true, false)] mailer: bool) {
    let generator = run_generator(mailer);

    let content =
        std::fs::read_to_string(generator.path("src/lib.rs")).expect("could not open file");

    if mailer {
        assertion::string::assert_line_regex(&content, "(?m)^pub mod mailers;$");
    } else {
        assertion::string::assert_str_not_exists(&content, "pub mod mailers;;");
    }
}
