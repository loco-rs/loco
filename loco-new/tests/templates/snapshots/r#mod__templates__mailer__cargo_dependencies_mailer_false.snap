---
source: loco-new/tests/templates/mailer.rs
expression: "content.get(\"dependencies\").unwrap()"
---
{ async-trait = "0.1.74", axum = "0.7.5", serde_json = "1", tracing = "0.1.40", loco-rs = { workspace = true }, serde = { features = ["derive"], version = "1" }, tokio = { default-features = false, features = ["rt-multi-thread"], version = "1.33.0" }, tracing-subscriber = { features = ["env-filter", "json"], version = "0.3.17" } }
